klippy/extras/palette2.py

```diff
@@ -0,0 +1,610 @@
+            "PALETTE_CUT", self.cmd_Cut, desc=self.cmd_Cut_Help)
+        self.gcode.register_command(
+            "PALETTE_SMART_LOAD",
+            self.cmd_Smart_Load,
+            desc=self.cmd_Smart_Load_Help)
+        self.serial = None
+        self.serial_port = config.get("serial")
+        if not self.serial_port:
+            raise config.error("Invalid serial port specific for Palette 2")
+        self.baud = config.getint("baud", default=250000)
+        self.feedrate_splice = config.getfloat(
+            "feedrate_splice", 0.8, minval=0., maxval=1.)
+        self.feedrate_normal = config.getfloat(
+            "feedrate_normal", 1.0, minval=0., maxval=1.)
+        self.auto_load_speed = config.getint("auto_load_speed", 2)
+
+        # Omega code matchers
+        self.omega_header = [None] * 9
+        omega_handlers = ["O" + str(i) for i in range(33)]
+        for cmd in omega_handlers:
+            func = getattr(self, 'cmd_' + cmd, None)
+            desc = getattr(self, 'cmd_' + cmd + '_help', None)
+            if func:
+                self.gcode.register_command(cmd, func, desc=desc)
+            else:
+                self.gcode.register_command(cmd, self.cmd_OmegaDefault)
+
+        self.pause_resume = None
+        self.virtual_sdcard = None
+        self._reset()
+
+        self.read_timer = None
+        self.read_buffer = ""
+        self.write_timer = None
+        self.write_queue = Queue()
+        self.heartbeat = None
+        self.last_heartbeat_send = 0
+
+        self.is_printing = False
+        self.smart_load_timer = None
+
+    def _reset(self):
+        self.files = []
+        self.is_setup_complete = False
+        self.is_splicing = False
+        self.is_loading = False
+        self.remaining_load_length = None
+        self.omega_algorithms = []
+        self.omega_algorithms_counter = 0
+        self.omega_splices = []
+        self.omega_splices_counter = 0
+        self.omega_pings = []
+        self.omega_pongs = []
+        self.omega_current_ping = None
+        self.omega_header = [None] * 9
+        self.omega_header_counter = 0
+        self.omega_last_command = ""
+        self.omega_drivers = []
+
+    def _check_P2(self, gcmd=None):
+        if self.serial:
+            return True
+        if gcmd:
+            gcmd.respond_info(INFO_NOT_CONNECTED)
+        return False
+
+    def handle_ready(self):
+        try:
+            self.virtual_sdcard = self.printer.lookup_object("virtual_sdcard")
+        except BaseException:
+            raise self.printer.config_error(
+                "Palette 2 requires [virtual_sdcard] to work,"
+                " please add it to your config!")
+
+        try:
+            self.pause_resume = self.printer.lookup_object("pause_resume")
+        except BaseException:
+            raise self.printer.config_error(
+                "Palette 2 requires [pause_resume] to work,"
+                " please add it to your config!")
```

Comment:
It's not valid to raise an error from the `handle_ready` event - at that point we're noting the printer is ready and we can't "not ready" it.  You could use `handle_connect` instead, but it's actually probably simpler to call `printer.load_object()` in the constructor instead - that will cause Klipper to automatically load the given modules.

Reply:
Ok, my understanding from the documentation was that we couldn't rely on load order of modules... though I guess I misinterpreted that to mean something other than what it did. I'll make this change and test shortly.